serial: uart_native_pty: IRQ support

tpambor opened the pull request:
Add support for the interrupt-driven API to uart_native_pty.

Interrupts are emulated using a polling thread.

aescolar reviewed (changes requested):
Thanks @tpambor :)
A few comments below

tpambor commented on drivers/serial/uart_native_pty.c:
```diff
@@ -403,6 +445,179 @@ static int np_uart_rx_disable(const struct device *dev)
+
+	while (1) {
+		if (data->irq.rx_enabled) {
+			int ret = np_uart_stdin_poll_in_bottom(data->in_fd, NULL, 0);
```
The async implementation uses only `np_uart_stdin_poll_in_bottom` as well? I was looking at the history, but I couldn't figure out why there are separate `np_uart_stdin_poll_in` and `np_uart_pty_poll_in`. Do you know if this is still necessary?

aescolar replied:
I took a look again now, one clear difference is the explicit handling of a disconnected pipe in the input.
This was actually what I had recalled when commenting this https://github.com/zephyrproject-rtos/zephyr/pull/93957#discussion_r2251188414 
i.e. I tend to recall explicitly adding support for a disconnecting input pipe needing extra care as it was a "normal" usecase for testing.

Maybe the easiest thing is to refactor a little bit the `np_uart_poll_in` `np_uart_stdin_poll_in` `np_uart_pty_poll_in` trio, something like:
```diff
diff --git a/drivers/serial/uart_native_pty.c b/drivers/serial/uart_native_pty.c
index 84dacf5c253..b238e0039cb 100644
--- a/drivers/serial/uart_native_pty.c
+++ b/drivers/serial/uart_native_pty.c
@@ -192,59 +192,50 @@ static void np_uart_poll_out(const struct device *dev, unsigned char out_char)
 }
 
 /**
- * @brief Poll the device for input.
+ * @brief Poll the device for up to len input characters
  *
  * @param dev UART device structure.
  * @param p_char Pointer to character.
  *
- * @retval 0 If a character arrived and was stored in p_char
+ * @retval > 0 If a character arrived and was stored in p_char
+ * @retval == 0 If a character arrived but len was 0
  * @retval -1 If no character was available to read
  */
-static int np_uart_stdin_poll_in(const struct device *dev, unsigned char *p_char)
+static int np_uart_poll_in_n(const struct device *dev, unsigned char *p_char, int len)
 {
+       int rc = -1;
        int in_f = ((struct native_pty_status *)dev->data)->in_fd;
-       static bool disconnected;
-       int rc;
-
-       if (disconnected == true) {
-               return -1;
-       }
 
-       rc = np_uart_stdin_poll_in_bottom(in_f, p_char, 1);
-       if (rc == -2) {
-               disconnected = true;
-       }
-       return rc == 1 ? 0 : -1;
-}
+       if (((struct native_pty_status *)dev->data)->on_stdinout) {
+               static bool disconnected;
 
-/**
- * @brief Poll the device for input.
- *
- * @param dev UART device structure.
- * @param p_char Pointer to character.
- *
- * @retval 0 If a character arrived and was stored in p_char
- * @retval -1 If no character was available to read
- */
-static int np_uart_pty_poll_in(const struct device *dev, unsigned char *p_char)
-{
-       int n = -1;
-       int in_f = ((struct native_pty_status *)dev->data)->in_fd;
+               if (disconnected == true) {
+                       return -1;
+               }
 
-       n = nsi_host_read(in_f, p_char, 1);
-       if (n == -1) {
+               rc = np_uart_stdin_poll_in_bottom(in_f, p_char, len);
+               if (rc == -2) {
+                       disconnected = true;
+                       return -1;
+               }
+               return rc;
+       } else {
+               rc = nsi_host_read(in_f, p_char, len);
+               if (rc > 0) {
+                       return rc;
+               }
                return -1;
        }
-       return 0;
 }
 
 static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
 {
-       if (((struct native_pty_status *)dev->data)->on_stdinout) {
-               return np_uart_stdin_poll_in(dev, p_char);
-       } else {
-               return np_uart_pty_poll_in(dev, p_char);
+       int ret = np_uart_poll_in_n(dev, p_char, 1);
+
+       if (ret == -1) {
+               return -1;
        }
+       return 0;
 }
 
 #ifdef CONFIG_UART_ASYNC_API
```

tpambor replied:
I refactored this similar to your suggestion and used `np_uart_poll_in_n()` in the new code.

tpambor commented on boards/native/native_sim/doc/index.rst:
```diff
@@ -551,8 +551,8 @@ option ``-attach_uart_cmd=<"cmd">``, or for each individual UART with
-This driver supports poll mode or async mode with :kconfig:option:`CONFIG_UART_ASYNC_API`.
-Interrupt mode is not supported.
+This driver supports poll mode, interrupt mode and async mode with
+:kconfig:option:`CONFIG_UART_ASYNC_API`.
```
I agree. Fixed.

tpambor commented on drivers/serial/uart_native_pty.c:
```diff
@@ -403,6 +445,179 @@ static int np_uart_rx_disable(const struct device *dev)
+			} else if (ret == -1) {
+				k_sleep(K_MSEC(1));
+			} else {
+				WARN("Poll returned error %d\n", ret);
```
This now uses `np_uart_poll_in_n`, so ret will be -1 if disconnected and the thread keeps sleeping and no more warnings are generated for this situation.

tpambor commented on drivers/serial/uart_native_pty.c:
```diff
@@ -403,6 +445,179 @@ static int np_uart_rx_disable(const struct device *dev)
+{
+	struct native_pty_status *data = dev->data;
+
+	return nsi_host_write(data->out_fd, (const void *)tx_data, size);
```
I changed this to use `np_uart_poll_out`. I think this way all cases should be covered.

tpambor commented on drivers/serial/uart_native_pty.c:
```diff
@@ -403,6 +445,179 @@ static int np_uart_rx_disable(const struct device *dev)
+{
+	struct native_pty_status *data = dev->data;
+
+	return np_uart_stdin_poll_in_bottom(data->in_fd, rx_data, size);
```
Uses now `np_uart_poll_in_n()`.

tpambor commented on drivers/serial/uart_native_pty.c:
```diff
@@ -403,6 +445,179 @@ static int np_uart_rx_disable(const struct device *dev)
+{
+	struct native_pty_status *data = dev->data;
+
+	if (data->irq.rx_enabled && np_uart_stdin_poll_in_bottom(data->in_fd, NULL, 0) == 0) {
```
Uses now `np_uart_poll_in_n()`.

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -337,8 +364,7 @@ static void native_pty_uart_async_poll_function(void *arg1, void *arg2, void *ar
 	while (data->async.rx_len) {
-		rc = np_uart_stdin_poll_in_bottom(data->in_fd, data->async.rx_buf,
-						  data->async.rx_len);
+		rc = np_uart_poll_in_n(dev, data->async.rx_buf, data->async.rx_len);
```
thank you :)

aescolar reviewed (changes requested):
Thanks @tpambor

tpambor commented on drivers/serial/uart_native_pty_bottom.c:
```diff
@@ -30,15 +30,19 @@
  * @retval -2 if the stdin is disconnected
  */
-int np_uart_stdin_poll_in_bottom(int in_f, unsigned char *p_char, int len)
+int np_uart_poll_in_bottom(int in_f, unsigned char *p_char, int len)
```
I oversaw this. This should be fixed now.

tpambor committed:
samples: sensor_shell: Wait for prompt before sending command

Wait for prompt before sending command to ensure shell is ready.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor committed:
serial: uart_native_pty: IRQ support

Add support for the interrupt-driven API. Interrupts are
emulated using a polling thread.

Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

tpambor commented:
Rebased on main to fix merge conflict in `boards/native/native_sim/doc/index.rst`.

tpambor commented:
> @tpambor I can see that in this case I just put above, `feof()` never claims the input is finished, while `select` keeps saying stdin is ready and `read` keeps returning 0 (eof). Meaning for stdin the logic that assumes that if select != 0 there is data is not holding, so first it appears there is data, and when the irq callback tries to read data there is no data. Leading to the infinite loop.

I fixed the end of file detection. I think `feof()` only works when characters a read using `fgetc()` etc. and not when using `read`. Your example no longer loops infinetly, but I still don't get output. I will investigate further.

aescolar reviewed (commented):
I think overall this is ok.
Let's wait a couple of hours to see what @jakub-uC says about
https://discord.com/channels/720317445772017664/937768560137220155/1404782521719328798
As I think we do not need to do anything here to work around that shell behavior
